File: src/uva/Leetcode00003.java
import java.util.*;

public class Leetcode00003 {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> mapIndexByCharacter = new HashMap<>();
        int max = 0;
        for (int i = 0;i < s.length();i++) {
            Integer index = mapIndexByCharacter.get(s.charAt(i));
            if (index != null) {
                i = index;
                mapIndexByCharacter.clear();
            } else {
                mapIndexByCharacter.put(s.charAt(i), i);
                max = Math.max(max, mapIndexByCharacter.size());
            }
        }
        return max;
    }
}

File: src/uva/Leetcode00003.swift
extension String {
    subscript (i: Int) -> Character {
        return self[self.startIndex.advancedBy(i)]
    }
    subscript (r: Range<Int>) -> String {
        return String(Array(arrayLiteral:self)[r.startIndex...r.endIndex])
    }
    var ascii: Int {
        return Int(self.unicodeScalars.first!.value)
    }
}
class Solution {
    func lengthOfLongestSubstring(s: String) -> Int {
        var mapIndexByChar = Array<Int?>(count: 256, repeatedValue: nil), i = 0, j = 0, l = 0
        let letters = s.characters.map { String($0) }
        while i < letters.count {
            var v = letters[i].ascii
            if let index = mapIndexByChar[v] {
                 j = max(j, index + 1)
            }
            mapIndexByChar[v] = i
            l = max(l, i - j + 1)
            i = i + 1
        }
        return l
    }
}
